stripeSdk has not been initialized on Android

I followed the example in order to do payments with webook and I believe I have not missed anything.
When I call confirmPaymentMethod: 
```dart
await Stripe.instance.confirmPaymentMethod(
    client_secret,
    PaymentMethodParams.card(),
);
```
I get this error:

<details>
<summary>Log error</summary>

```E/MethodChannel#flutter.stripe/payments( 9694): Failed to handle method call
E/MethodChannel#flutter.stripe/payments( 9694): kotlin.UninitializedPropertyAccessException: lateinit property stripeSdk has not been initialized
E/MethodChannel#flutter.stripe/payments( 9694):         at com.flutter.stripe.StripeAndroidPlugin.onMethodCall(StripeAndroidPlugin.kt:44)
E/MethodChannel#flutter.stripe/payments( 9694):         at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler.onMessage(MethodChannel.java:233)
E/MethodChannel#flutter.stripe/payments( 9694):         at io.flutter.embedding.engine.dart.DartMessenger.handleMessageFromDart(DartMessenger.java:85)
E/MethodChannel#flutter.stripe/payments( 9694):         at io.flutter.embedding.engine.FlutterJNI.handlePlatformMessage(FlutterJNI.java:818)
E/MethodChannel#flutter.stripe/payments( 9694):         at android.os.MessageQueue.nativePollOnce(Native Method)
E/MethodChannel#flutter.stripe/payments( 9694):         at android.os.MessageQueue.next(MessageQueue.java:363)
E/MethodChannel#flutter.stripe/payments( 9694):         at android.os.Looper.loop(Looper.java:173)
E/MethodChannel#flutter.stripe/payments( 9694):         at android.app.ActivityThread.main(ActivityThread.java:8178)
E/MethodChannel#flutter.stripe/payments( 9694):         at java.lang.reflect.Method.invoke(Native Method)
E/MethodChannel#flutter.stripe/payments( 9694):         at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:513)
E/MethodChannel#flutter.stripe/payments( 9694):         at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1101)
E/flutter ( 9694): [ERROR:flutter/lib/ui/ui_dart_state.cc(199)] Unhandled Exception: PlatformException(error, lateinit property stripeSdk has not been initialized, null, kotlin.UninitializedPropertyAccessException: lateinit property stripeSdk has not been initialized
E/flutter ( 9694):      at com.flutter.stripe.StripeAndroidPlugin.onMethodCall(StripeAndroidPlugin.kt:44)
E/flutter ( 9694):      at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler.onMessage(MethodChannel.java:233)
E/flutter ( 9694):      at io.flutter.embedding.engine.dart.DartMessenger.handleMessageFromDart(DartMessenger.java:85)
E/flutter ( 9694):      at io.flutter.embedding.engine.FlutterJNI.handlePlatformMessage(FlutterJNI.java:818)
E/flutter ( 9694):      at android.os.MessageQueue.nativePollOnce(Native Method)
E/flutter ( 9694):      at android.os.MessageQueue.next(MessageQueue.java:363)
E/flutter ( 9694):      at android.os.Looper.loop(Looper.java:173)
E/flutter ( 9694):      at android.app.ActivityThread.main(ActivityThread.java:8178)
E/flutter ( 9694):      at java.lang.reflect.Method.invoke(Native Method)
E/flutter ( 9694):      at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:513)
E/flutter ( 9694):      at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1101)
E/flutter ( 9694): )
E/flutter ( 9694): #0      JSONMethodCodec.decodeEnvelope (package:flutter/src/services/message_codecs.dart:164:7)
E/flutter ( 9694): #1      MethodChannel._invokeMethod (package:flutter/src/services/platform_channel.dart:158:18)
E/flutter ( 9694): <asynchronous suspension>
E/flutter ( 9694): #2      MethodChannelStripe.initialise (package:stripe_platform_interface/src/method_channel_stripe.dart:36:5)
E/flutter ( 9694): <asynchronous suspension>
E/flutter ( 9694): #3      Stripe._initialise (package:flutter_stripe/src/stripe.dart:315:5)
E/flutter ( 9694): <asynchronous suspension>
E/flutter ( 9694): #4      Stripe.confirmPaymentMethod (package:flutter_stripe/src/stripe.dart:186:5)
E/flutter ( 9694): <asynchronous suspension>
```
</details>


The thing is that the example works, my app doesn't.
I've call also in my main.dart before runApp and public key already configured: 
```dart 
WidgetsFlutterBinding.ensureInitialized();
```

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, 2.2.0, on Linux, locale en_GB.UTF-8)
    • Flutter version 2.2.0 at /home/anmentone/lib/flutter
    • Framework revision b22742018b (10 days ago), 2021-05-14 19:12:57 -0700
    • Engine revision a9d88a4d18
    • Dart version 2.13.0

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /home/anmentone/Android/Sdk
    • Platform android-30, build-tools 30.0.3
    • Java binary at: /usr/local/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.8+0-b944-P17168821)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Android Studio (version 4.2)
    • Android Studio at /usr/local/android-studio
    • Flutter plugin version 56.0.2
    • Dart plugin version 202.8488
    • Java version OpenJDK Runtime Environment (build 11.0.8+0-b944-P17168821)

[✓] IntelliJ IDEA Community Edition (version 2021.1)
    • IntelliJ at /usr/local/idea-IC-211.7142.45
    • Flutter plugin version 56.0.5
    • Dart plugin version 211.7233

[✓] VS Code (version 1.56.2)
    • VS Code at /usr/share/code
    • Flutter extension version 3.22.0

[✓] Connected device (2 available)
    • POT LX1 (mobile) • 39V4C19A18025168 • android-arm64  • Android 10 (API 29)
    • Chrome (web)     • chrome           • web-javascript • Google Chrome 90.0.4430.212

• No issues found!
```
</details>

 - Package version: 1.0.1+2

Thanks for your help!

Hi guys, I'm back here writing again.
I'm developing a plugin for flutter to insert card payment functionality inside my flutter app for mypos using mypos SDK (myPOS-Smart-SDK). 
My app is about "digital commercial documents" (receipt) and is used by vendors instead of a physical cash register.  
I only use stripe for subscription renewal, which is required to use some services in the app.
So these two plugins are always used separately but I'm having some unexpected issues.
The functioning of the app is like a simple cash register, when there is a sale there is a request for payment and the customer can decide to pay by card using the integrated card reader of mypos. The payment handling is done with the plugin I'm developing that consumes the mypos API internally.
The strange thing is that when the payment finishes and **mypos close its activity, the app crashes throwing me this exception**:

<details> 
<summary>kotlin.UninitializedPropertyAccessException: lateinit property stripe has not been initialized</summary>

```
E/AndroidRuntime(25099): 	at android.os.Handler.dispatchMessage(Handler.java:102)
E/AndroidRuntime(25099): 	at android.os.Looper.loop(Looper.java:135)
E/AndroidRuntime(25099): 	at android.app.ActivityThread.main(ActivityThread.java:5254)
E/AndroidRuntime(25099): 	at java.lang.reflect.Method.invoke(Native Method)
E/AndroidRuntime(25099): 	at java.lang.reflect.Method.invoke(Method.java:372)
E/AndroidRuntime(25099): 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:902)
E/AndroidRuntime(25099): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:697)
E/AndroidRuntime(25099): Caused by: kotlin.UninitializedPropertyAccessException: lateinit property stripe has not been initialized
E/AndroidRuntime(25099): 	at com.reactnativestripesdk.StripeSdkModule.access$getStripe$p(StripeSdkModule.kt:20)
E/AndroidRuntime(25099): 	at com.reactnativestripesdk.StripeSdkModule$mActivityEventListener$1.onActivityResult(StripeSdkModule.kt:45)
E/AndroidRuntime(25099): 	at com.facebook.react.bridge.BaseActivityEventListener.onActivityResult(BaseActivityEventListener.kt:15)
E/AndroidRuntime(25099): 	at io.flutter.embedding.engine.FlutterEngineConnectionRegistry$FlutterEngineActivityPluginBinding.onActivityResult(FlutterEngineConnectionRegistry.java:739)
E/AndroidRuntime(25099): 	at io.flutter.embedding.engine.FlutterEngineConnectionRegistry.onActivityResult(FlutterEngineConnectionRegistry.java:426)
E/AndroidRuntime(25099): 	at io.flutter.embedding.android.FlutterActivityAndFragmentDelegate.onActivityResult(FlutterActivityAndFragmentDelegate.java:677)
E/AndroidRuntime(25099): 	at io.flutter.embedding.android.FlutterFragment.onActivityResult(FlutterFragment.java:769)
E/AndroidRuntime(25099): 	at io.flutter.embedding.android.FlutterFragmentActivity.onActivityResult(FlutterFragmentActivity.java:509)
E/AndroidRuntime(25099): 	at android.app.Activity.dispatchActivityResult(Activity.java:6235)
E/AndroidRuntime(25099): 	at android.app.ActivityThread.deliverResults(ActivityThread.java:3570)
E/AndroidRuntime(25099): 	... 10 more
I/Process (25099): Sending signal. PID: 25099 SIG: 9
```

</details>

This left me surprised as **I am not using Stripe** in this segment.

Seeing the message error, I actually tried using stripe before I could make a payment with the mypos API, and it went like this...
After making the payment **I get a black screen but the app doesn't crash** and this is the error message:


<details> 
<summary>Activity com.anmentone.app.MainActivity has leaked IntentReceiver</summary>

```
E/ActivityThread(23524): android.app.IntentReceiverLeaked: Activity com.anmentone.app.MainActivity has leaked IntentReceiver com.reactnativestripesdk.StripeSdkModule$mPaymentSheetReceiver$1@3d98bd7f that was originally registered here. Are you missing a call to unregisterReceiver()?
E/ActivityThread(23524): 	at android.app.LoadedApk$ReceiverDispatcher.<init>(LoadedApk.java:903)
E/ActivityThread(23524): 	at android.app.LoadedApk.getReceiverDispatcher(LoadedApk.java:704)
E/ActivityThread(23524): 	at android.app.ContextImpl.registerReceiverInternal(ContextImpl.java:1650)
E/ActivityThread(23524): 	at android.app.ContextImpl.registerReceiver(ContextImpl.java:1630)
E/ActivityThread(23524): 	at android.app.ContextImpl.registerReceiver(ContextImpl.java:1624)
E/ActivityThread(23524): 	at android.content.ContextWrapper.registerReceiver(ContextWrapper.java:489)
E/ActivityThread(23524): 	at com.facebook.react.bridge.ReactComponentActivityWrapper.registerReceiver(ReactComponentActivityWrapper.java:39)
E/ActivityThread(23524): 	at com.reactnativestripesdk.StripeSdkModule.initialise(StripeSdkModule.kt:240)
E/ActivityThread(23524): 	at com.flutter.stripe.StripeAndroidPlugin.onMethodCall(StripeAndroidPlugin.kt:44)
E/ActivityThread(23524): 	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler.onMessage(MethodChannel.java:233)
E/ActivityThread(23524): 	at io.flutter.embedding.engine.dart.DartMessenger.handleMessageFromDart(DartMessenger.java:85)
E/ActivityThread(23524): 	at io.flutter.embedding.engine.FlutterJNI.handlePlatformMessage(FlutterJNI.java:818)
E/ActivityThread(23524): 	at android.os.MessageQueue.nativePollOnce(Native Method)
E/ActivityThread(23524): 	at android.os.MessageQueue.next(MessageQueue.java:143)
E/ActivityThread(23524): 	at android.os.Looper.loop(Looper.java:122)
E/ActivityThread(23524): 	at android.app.ActivityThread.main(ActivityThread.java:5254)
E/ActivityThread(23524): 	at java.lang.reflect.Method.invoke(Native Method)
E/ActivityThread(23524): 	at java.lang.reflect.Method.invoke(Method.java:372)
E/ActivityThread(23524): 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:902)
E/ActivityThread(23524): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:697)
```

</details>

<details>
<summary>Device info</summary>

myPOS Smart N5
Operating system | Android 5.1
-- | --
Processor | 4 Core CPU + Secure CPU
Memory | 8GB FLASH, 1GB RAM
Display | 5.5’ (720 x 1280) Colorful LCD capacitive touchscreen with signature capture functionality
Printer | High-Speed Thermal PrinterPaper Roll Width/Diameter: 58mm/40mm
Communications | 4G LTE/ Wi-Fi / Bluetooth
Camera | 5MP - rear -facing
Data Capture | 1D Barcode scanner (optional)2MP camera -front-facing (optional)
GPS Tracking | GPS / BEIDOU / GLONASS
Card Readers | Magstripe: Triple track, high coercivity, bi-directionalSmart Card: EMV &amp; PBOC approvedContactless : NFC 13.56MHz, support ISO14443, Type A/B/C, MifareCard (optional)
Card Slots | PSAM: 2SIM: 1SD: 1
Battery | Li-ion 4300mAhInput : 100-240V, 50-60HZOutput: 5V/2A
Multi-media | Video, Audio
Dimensions | L × W × H = 186mm × 82mm × 64mm , 417g
Device Ports | 1 Micro-USB OTG
Environmental | Operating Temperature: 0°C ̴ 50°C (32°F ̴ 122°F)&nbsp;&nbsp;Storage Temperature: -20°C ̴ 60°C (-4°F ̴ 158°F)&nbsp;Relative Humidity: 5% ̴ 95% (Non-Condensing)
Optional Base | Charging Base / Wi-Fi Base: Wi-Fi, integrated charging
Certifications | PCI 4.X, EMV, PayPass, PayWave certified
Languages | English, German, Bulgarian, Czech, Greek, Spanish, French, Croatian, Italian, Dutch, Portugal, Romanian, Swedish, Hungarian, Icelandic, Latvian, Polish, Slovenian, Norwegian
</details>


I emphasize that Stripe has nothing to do with these payments. I use it elsewhere to be able to renew the subscription which allows you to use premium features.

I'm trying to go deeper but I can't understand the problem. If you have any ideas I would be very grateful.
Thank you!

--- a/packages/stripe_android/android/src/main/kotlin/com/facebook/react/bridge/BaseActivityEventListener.kt
+++ b/packages/stripe_android/android/src/main/kotlin/com/facebook/react/bridge/BaseActivityEventListener.kt
@@ -2,17 +2,20 @@ package com.facebook.react.bridge
 
 import android.app.Activity
 import android.content.Intent
-import androidx.annotation.NonNull
-import com.stripe.android.Stripe
+import com.flutter.stripe.StripeAndroidPlugin
 import io.flutter.plugin.common.PluginRegistry.ActivityResultListener
 import java.lang.ref.WeakReference
 
 abstract class BaseActivityEventListener : ActivityEventListener, ActivityResultListener {
 
+    var pluginInstance: StripeAndroidPlugin? = null
+
     lateinit var activity: WeakReference<Activity>
 
     override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?): Boolean {
-        onActivityResult(activity.get(), requestCode, resultCode, data)
+        if (pluginInstance != null && pluginInstance?.isInitialized == true) {
+            onActivityResult(activity.get(), requestCode, resultCode, data)
+        }
         return false
     }
 }

--- a/packages/stripe_android/android/src/main/kotlin/com/facebook/react/bridge/Promise.java
+++ b/packages/stripe_android/android/src/main/kotlin/com/facebook/react/bridge/Promise.java
@@ -1,5 +1,8 @@
 package com.facebook.react.bridge;
 
+import android.os.Handler;
+import android.os.Looper;
+
 import io.flutter.plugin.common.MethodChannel;
 
 /**
@@ -7,18 +10,30 @@ import io.flutter.plugin.common.MethodChannel;
  */
 public class Promise {
 
-    final MethodChannel.Result result;
+    final MethodChannel.Result methodResult;
+    private final Handler mainHandler;
 
     public Promise(MethodChannel.Result result) {
-        this.result = result;
+        this.methodResult = result;
+        mainHandler = new Handler(Looper.getMainLooper());
     }
 
-    public void resolve(Object result) {
-        this.result.success(result);
+    public void resolve(final Object result) {
+        mainHandler.post(new Runnable() {
+            @Override
+            public void run() {
+                methodResult.success(result);
+            }
+        });
     }
 
-    public void reject(String errorCode, String message) {
-        this.result.error(errorCode, message, null);
+    public void reject(final String errorCode, final String message) {
+        mainHandler.post(new Runnable() {
+            @Override
+            public void run() {
+                methodResult.error(errorCode, message, null);
+            }
+        });
     }
 
     public void reject(String errorCode) {

--- a/packages/stripe_android/android/src/main/kotlin/com/facebook/react/bridge/ReactApplicationContext.java
+++ b/packages/stripe_android/android/src/main/kotlin/com/facebook/react/bridge/ReactApplicationContext.java
@@ -1,8 +1,9 @@
 package com.facebook.react.bridge;
 
 import android.app.Activity;
-import android.content.Context;
 import android.content.ContextWrapper;
+
+import com.flutter.stripe.StripeAndroidPlugin;
 
 import org.jetbrains.annotations.NotNull;
 
@@ -13,14 +14,17 @@ import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
 public class ReactApplicationContext extends ContextWrapper {
 
     private final ActivityPluginBinding binding;
+    private final StripeAndroidPlugin pluginInstance;
 
-    public ReactApplicationContext(ActivityPluginBinding binding) {
+    public ReactApplicationContext(ActivityPluginBinding binding, StripeAndroidPlugin pluginInstance) {
         super(binding.getActivity());
         this.binding = binding;
+        this.pluginInstance = pluginInstance;
     }
 
     public void addActivityEventListener(@NotNull BaseActivityEventListener activityEventListener) {
         activityEventListener.activity = new WeakReference(binding.getActivity());
+        activityEventListener.setPluginInstance(pluginInstance);
         binding.addActivityResultListener(activityEventListener);
     }
 

--- a/packages/stripe_android/android/src/main/kotlin/com/facebook/react/bridge/ReactComponentActivityWrapper.java
+++ b/packages/stripe_android/android/src/main/kotlin/com/facebook/react/bridge/ReactComponentActivityWrapper.java
@@ -1,18 +1,13 @@
 package com.facebook.react.bridge;
 
 import android.annotation.SuppressLint;
-import android.app.Activity;
 import android.content.BroadcastReceiver;
 import android.content.Intent;
 import android.content.IntentFilter;
 
-import androidx.annotation.NonNull;
-import androidx.appcompat.app.AppCompatActivity;
-import androidx.core.app.ComponentActivity;
-import androidx.fragment.app.FragmentManager;
-
 import org.jetbrains.annotations.NotNull;
 
+import java.util.HashMap;
 import java.util.HashSet;
 import java.util.Set;
 
@@ -23,7 +18,7 @@ public class ReactComponentActivityWrapper {
 
     private final FlutterFragmentActivity activity;
 
-    private final Set<String> registeredIntentFilterActions = new HashSet<>();
+    private final HashMap<String, BroadcastReceiver> registeredIntentFilters = new HashMap<>();
 
     public ReactComponentActivityWrapper(FlutterFragmentActivity activity) {
         this.activity = activity;
@@ -34,12 +29,23 @@ public class ReactComponentActivityWrapper {
     }
 
     public Intent registerReceiver(@NotNull BroadcastReceiver receiver, @NotNull IntentFilter intentFilter) {
-        if (!registeredIntentFilterActions.contains(intentFilter.getAction(0))) {
-            registeredIntentFilterActions.add(intentFilter.getAction(0));
+        if (!registeredIntentFilters.containsKey(intentFilter.getAction(0))) {
+            registeredIntentFilters.put(intentFilter.getAction(0), receiver);
             return activity.registerReceiver(receiver, intentFilter);
-        } else {
-            System.out.println("HI");
         }
         return null;
+    }
+
+    public void unregisterReceivers() {
+        final Set<String> keys = new HashSet<>(registeredIntentFilters.keySet());
+        for (String action : keys) {
+            BroadcastReceiver broadcastReceiver = registeredIntentFilters.remove(action);
+            try {
+                activity.unregisterReceiver(broadcastReceiver);
+            } catch (IllegalArgumentException e) {
+                // no-op
+            }
+        }
+        assert registeredIntentFilters.isEmpty();
     }
 }

--- a/packages/stripe_android/android/src/main/kotlin/com/facebook/react/bridge/ReactContextBaseJavaModule.java
+++ b/packages/stripe_android/android/src/main/kotlin/com/facebook/react/bridge/ReactContextBaseJavaModule.java
@@ -1,16 +1,12 @@
 package com.facebook.react.bridge;
 
-import android.annotation.SuppressLint;
 import android.app.Activity;
 import android.content.Context;
 import android.content.Intent;
 
-import androidx.core.app.ComponentActivity;
-
 import java.util.ArrayList;
 
 import io.flutter.embedding.android.FlutterFragmentActivity;
-import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
 import io.flutter.plugin.common.PluginRegistry;
 
 public class ReactContextBaseJavaModule implements PluginRegistry.ActivityResultListener {
@@ -31,16 +27,8 @@ public class ReactContextBaseJavaModule implements PluginRegistry.ActivityResult
         return activity;
     }
 
-    protected ReactComponentActivityWrapper getCurrentActivity() {
+    public ReactComponentActivityWrapper getCurrentActivity() {
         return activityWrapper;
-    }
-
-    protected void addActivityEventListener(ActivityEventListener listener) {
-        eventListeners.add(listener);
-    }
-
-    protected void removeActivityEventListener(ActivityEventListener listener) {
-        eventListeners.remove(listener);
     }
 
     public String getName() {

--- a/packages/stripe_android/android/src/main/kotlin/com/flutter/stripe/StripeAndroidPlugin.kt
+++ b/packages/stripe_android/android/src/main/kotlin/com/flutter/stripe/StripeAndroidPlugin.kt
@@ -6,6 +6,7 @@ import com.facebook.react.bridge.ReactApplicationContext
 import com.facebook.react.bridge.ReadableMap
 import com.reactnativestripesdk.StripeSdkCardViewManager
 import com.reactnativestripesdk.StripeSdkModule
+import io.flutter.embedding.android.FlutterFragmentActivity
 import io.flutter.embedding.engine.plugins.FlutterPlugin
 import io.flutter.embedding.engine.plugins.activity.ActivityAware
 import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
@@ -24,7 +25,11 @@ class StripeAndroidPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
     /// This local reference serves to register the plugin with the Flutter Engine and unregister it
     /// when the Flutter Engine is detached from the Activity
     private lateinit var channel: MethodChannel
-    private lateinit var stripeSdk: StripeSdkModule
+    lateinit var stripeSdk: StripeSdkModule
+
+    // set to true when the plugin has initialized - this is used as an indicator to receive
+    // activity results
+    var isInitialized = false
 
     private val stripeSdkCardViewManager: StripeSdkCardViewManager by lazy {
         StripeSdkCardViewManager()
@@ -39,12 +44,21 @@ class StripeAndroidPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
     }
 
     override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
+        if (!this::stripeSdk.isInitialized) {
+            result.error(
+        "flutter_stripe initialization failed",
+        "The plugin failed to initialize. Are you using FlutterFragmentActivity? Please check the README: https://github.com/flutter-stripe/flutter_stripe#android",
+        null
+            )
+            return
+        }
         when (call.method) {
             "initialise" -> {
                 stripeSdk.initialise(
                         params = ReadableMap(call.arguments as JSONObject),
                         promise = Promise(result),
                 )
+                isInitialized = true
             }
             "createPaymentMethod" -> stripeSdk.createPaymentMethod(
                     data = call.requiredArgument("data"),
@@ -103,7 +117,11 @@ class StripeAndroidPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
     }
 
     override fun onAttachedToActivity(binding: ActivityPluginBinding) {
-        stripeSdk = StripeSdkModule(ReactApplicationContext(binding), stripeSdkCardViewManager)
+        if (binding.activity is FlutterFragmentActivity) {
+            stripeSdk = StripeSdkModule(ReactApplicationContext(binding, this), stripeSdkCardViewManager)
+        } else {
+            // no-op - will throw errors when method channel is called
+        }
     }
 
     override fun onDetachedFromActivityForConfigChanges() {
@@ -113,6 +131,9 @@ class StripeAndroidPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
     }
 
     override fun onDetachedFromActivity() {
+        if (this::stripeSdk.isInitialized) {
+            stripeSdk.currentActivity.unregisterReceivers()
+        }
     }
 }
 

--- a/packages/stripe_android/android/src/main/kotlin/com/reactnativestripesdk/StripeSdkModule.kt
+++ b/packages/stripe_android/android/src/main/kotlin/com/reactnativestripesdk/StripeSdkModule.kt
@@ -305,7 +305,7 @@ class StripeSdkModule(reactContext: ReactApplicationContext, cardFieldManager: S
   }
 
   private fun payWithFpx() {
-    AddPaymentMethodActivityStarter(currentActivity as AppCompatActivity)
+    AddPaymentMethodActivityStarter(currentActivity.activity as AppCompatActivity)
             .startForResult(AddPaymentMethodActivityStarter.Args.Builder()
                     .setPaymentMethodType(PaymentMethod.Type.Fpx)
                     .build()

